File: solution/930. Binary Subarrays With Sum/930_02.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return atMost(nums, goal) - atMost(nums, goal - 1);
    }
    
    int atMost(vector<int>& nums, int goal) {
        if (goal < 0) return 0;
        int res = 0, i = 0, n = nums.size();
        for (int j = 0; j < n; ++j) {
            goal -= nums[j];
            while (goal < 0)
                goal += nums[i++];
            res += j - i + 1;
        }
        return res;
    }
};

File: solution/930. Binary Subarrays With Sum/930_02.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        def atMost(val):
            if val < 0: return 0
            res = i = 0
            for j in range(len(nums)):
                val -= nums[j]
                while val < 0:
                    val += nums[i]
                    i += 1
                res += j - i + 1
            return res
        
        return atMost(goal) - atMost(goal - 1)
    
